File: Qin/Qin-ios/Qin-ios/Util/DatabaseHelper.swift
//
//  DatabaseHelper.swift
//  Qin-ios
//
//  Created by 杨璨 on 2020/1/6.
//  Copyright © 2020 canyang. All rights reserved.
//

import Foundation
import RealmSwift

public class DatabaseHelper {
    //Lean Cloud
    public static var LeancloudAppid = "N4v46EIBIAWtiOANE61Fe1no-gzGzoHsz";
    public static var LeancloudAppKey = "RCzPdQyEuPLaFhcPlxaKVb9P";
    public static var LeancloudAPIBaseURL = "https://n4v46eib.lc-cn-n1-shared.com";
    public static var LeancloudIDHeader = "X-LC-Id";
    public static var LeancloudKeyHeader = "X-LC-Key";
    public static var HttpContentTypeHeader = "Content-Type";
    public static var HttpContentTypeJSONUTF8 = "application/json; charset=utf-8";
    
    public static func LCSearch(searchURL: String, completionHandler: @escaping ([String:Any], Error?) -> Void) {
        let createCheckItemURL = URL(string: searchURL)
        let leancloudSession = URLSession.shared;
        var leancloudRequest = URLRequest(url: createCheckItemURL!);
        leancloudRequest.httpMethod = "GET"
        leancloudRequest.setValue(HttpContentTypeJSONUTF8, forHTTPHeaderField: HttpContentTypeHeader)
        leancloudRequest.setValue(LeancloudAppid, forHTTPHeaderField: LeancloudIDHeader)
        leancloudRequest.setValue(LeancloudAppKey, forHTTPHeaderField: LeancloudKeyHeader)
        leancloudSession.dataTask(with: leancloudRequest) { data, response, error in
            // Do something...
            do{
                if error != nil {
                    completionHandler([String:Any](), error)
                } else {
                    let json = try JSONSerialization.jsonObject(with: data!, options: []) as! [String:Any]
                    completionHandler(json, nil)
                }
            }catch{
                completionHandler([String:Any](), error)
            }
        }.resume()
    }
    
    public static func LCUpdateAdvertising(completionHandler: @escaping (Bool) -> Void){
        let url = DatabaseHelper.LeancloudAPIBaseURL + "/1.1/classes/Student/" + ApplicationHelper.CurrentUser.LCObjectID
        let createCheckItemURL = URL(string: url)
        let leancloudSession = URLSession.shared;
        var leancloudRequest = URLRequest(url: createCheckItemURL!);
        leancloudRequest.httpMethod = "PUT"
        leancloudRequest.setValue(HttpContentTypeJSONUTF8, forHTTPHeaderField: HttpContentTypeHeader)
        leancloudRequest.setValue(LeancloudAppid, forHTTPHeaderField: LeancloudIDHeader)
        leancloudRequest.setValue(LeancloudAppKey, forHTTPHeaderField: LeancloudKeyHeader)
        let checkJson = ["Advertising": "1"]
        let checkJSONData = try? JSONSerialization.data(withJSONObject: checkJson, options: [])
        leancloudSession.uploadTask(with: leancloudRequest, from: checkJSONData) { data, response, error in
            if let httpres = response as? HTTPURLResponse {
                completionHandler(httpres.statusCode == 200)
            } else {
                completionHandler(false)
            }
        }.resume()
    }

    public static func LCCheckAdvertising(value:String, completionHandler: @escaping (Bool) -> Void) {
        let url = DatabaseHelper.LeancloudAPIBaseURL + "/1.1/classes/Student/" + ApplicationHelper.CurrentUser.LCObjectID;
        DatabaseHelper.LCSearch(searchURL: url){ json, error in
            if let advertiding = json["Advertising"] as? String , error == nil {
                //print("id: \(ApplicationHelper.CurrentUser.SchoolID) adv: \(advertiding)  value: \(value)")
                completionHandler(advertiding == value)
            } else {
                print("false")
                completionHandler(false)
            }
        }
    }

    public static func LCUploadCheckLog(completionHandler: @escaping (Bool) -> Void) {
        var checkInRoom = Room()
        autoreleasepool {
            checkInRoom =  (try! Realm()).objects(Room.self).filter("ID = \(ApplicationHelper.CheckInRoomID)").first!
        }
        let url = DatabaseHelper.LeancloudAPIBaseURL + "/1.1/classes/CheckRecording"
        let createCheckItemURL = URL(string: url)
        let leancloudSession = URLSession.shared;
        var leancloudRequest = URLRequest(url: createCheckItemURL!);
        leancloudRequest.httpMethod = "POST"
        leancloudRequest.setValue(HttpContentTypeJSONUTF8, forHTTPHeaderField: HttpContentTypeHeader)
        leancloudRequest.setValue(LeancloudAppid, forHTTPHeaderField: LeancloudIDHeader)
        leancloudRequest.setValue(LeancloudAppKey, forHTTPHeaderField: LeancloudKeyHeader)
        let checkJson = ["StudentID"    : ApplicationHelper.CurrentUser.ID,
                         "RoomID"       : checkInRoom.ID]
        let checkJSONData = try? JSONSerialization.data(withJSONObject: checkJson, options: [])
        leancloudSession.uploadTask(with: leancloudRequest, from: checkJSONData) { data, response, error in
            if let httpres = response as? HTTPURLResponse {
                completionHandler(httpres.statusCode == 201)
            } else {
                completionHandler(false)
            }
        }.resume()
    }

}

File: Qin/Qin-Android/app/src/main/java/edu/bfa/ss/qin/Util/DatabaseHelper.java
package edu.bfa.ss.qin.Util;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;

import javax.net.ssl.HttpsURLConnection;

import io.realm.Realm;

public class DatabaseHelper {
    //Lean Cloud
    public static String LeancloudAppid = "Tf0m64H1aEhwItMDiMH87pD7-gzGzoHsz";
    public static String LeancloudAppKey = "SWhko62oywljuSCkqRnNdjiM";
    public static String LeancloudAPIBaseURL = "https://tf0m64h1.lc-cn-n1-shared.com";
    public static String LeancloudIDHeader = "X-LC-Id";
    public static String LeancloudKeyHeader = "X-LC-Key";
    public static String HttpContentTypeHeader = "Content-Type";
    public static String HttpContentTypeJSONUTF8 = "application/json; charset=utf-8";

    public static JSONObject LCSearch(String searchURL) throws IOException, JSONException {
        HttpsURLConnection connection = (HttpsURLConnection) (new URL(searchURL)).openConnection();
        connection.setRequestMethod("GET");
        connection.setRequestProperty(DatabaseHelper.LeancloudIDHeader, DatabaseHelper.LeancloudAppid);
        connection.setRequestProperty(DatabaseHelper.LeancloudKeyHeader, DatabaseHelper.LeancloudAppKey);
        connection.setRequestProperty(DatabaseHelper.HttpContentTypeHeader, DatabaseHelper.HttpContentTypeJSONUTF8);
        if (connection.getResponseCode() == 200) {
            return new JSONObject(new BufferedReader(new InputStreamReader(connection.getInputStream())).readLine());
        }
        throw new JSONException("获取失败");
    }

    public static boolean LCUpdateAdvertising(){
        try {
            String url = DatabaseHelper.LeancloudAPIBaseURL + "/1.1/classes/Student/" + ApplicationHelper.CurrentUser.LCObjectID;
            HttpsURLConnection connection = (HttpsURLConnection) (new URL(url)).openConnection();
            connection.setRequestMethod("PUT");
            connection.setRequestProperty(DatabaseHelper.LeancloudIDHeader, DatabaseHelper.LeancloudAppid);
            connection.setRequestProperty(DatabaseHelper.LeancloudKeyHeader, DatabaseHelper.LeancloudAppKey);
            connection.setRequestProperty(DatabaseHelper.HttpContentTypeHeader, DatabaseHelper.HttpContentTypeJSONUTF8);
            JSONObject jsonParam = new JSONObject();
            jsonParam.put("Advertising", "1");
            DataOutputStream os = new DataOutputStream(connection.getOutputStream());
            os.writeBytes(jsonParam.toString());
            os.flush();
            os.close();
            return connection.getResponseCode() == 200;
        } catch (Exception e) {
            return false;
        }
    }

    public static boolean LCCheckAdvertising(String value) {
        try {
            String url = DatabaseHelper.LeancloudAPIBaseURL + "/1.1/classes/Student/" + ApplicationHelper.CurrentUser.LCObjectID;
            JSONObject response = DatabaseHelper.LCSearch(url);
            return response.getString("Advertising").equals(value);
        } catch (Exception e) {
            return  false;
        }
    }

    public static boolean LCUploadCheckLog() {
        String url = LeancloudAPIBaseURL + "/1.1/classes/CheckRecording";
        Realm realm = Realm.getDefaultInstance();
        Room checkInRoom = realm.where(Room.class).equalTo("BLE", ApplicationHelper.CheckInRoomID).findFirst();
        try {
            HttpsURLConnection connection = (HttpsURLConnection) (new URL(url)).openConnection();
            connection.setRequestMethod("POST");
            connection.setRequestProperty(LeancloudIDHeader, LeancloudAppid);
            connection.setRequestProperty(LeancloudKeyHeader, LeancloudAppKey);
            connection.setRequestProperty(HttpContentTypeHeader, HttpContentTypeJSONUTF8);
            connection.setDoOutput(true);
            connection.setDoInput(true);
            JSONObject jsonParam = new JSONObject();
            jsonParam.put("StudentID", ApplicationHelper.CurrentUser.BLE);
            jsonParam.put("RoomID", checkInRoom.ID);
            DataOutputStream os = new DataOutputStream(connection.getOutputStream());
            os.writeBytes(jsonParam.toString());
            os.flush();
            os.close();
            return connection.getResponseCode() == 201;
        } catch (Exception e) {
          return false;
        } finally {
            realm.close();
            realm = null;
        }
    }
}
